src/main/scala/io/iohk/ethereum/vm/WorldStateProxy.scala

```diff
@@ -13,35 +13,39 @@ import io.iohk.ethereum.crypto.kec256
   * should be kept in memory and applied only after a transaction completes without errors. This does not forbid mutable
   * caches for DB retrieval operations.
   */
-trait WorldStateProxy {
-  type WS <: WorldStateProxy
-
-  def getAccount(address: Address): Option[Account]
-  def getCode(codeHash: ByteString): ByteString
-  def getStorage(storageRoot: ByteString): Storage
-
-  def saveAccount(address: Address, account: Account): WS
-  def saveCode(codeHash: ByteString, code: ByteString): WS
-  def saveStorage(storageRoot: ByteString, storage: Storage): WS
+trait WorldStateProxy[WS <: WorldStateProxy[WS, S], S <: Storage[S]] {
 
+  protected def getAccount(address: Address): Option[Account]
+  protected def saveAccount(address: Address, account: Account): WS
```

Comment:
if implementator needs these methods to return concrete types he can always override them such way (i.e specify subtype as a return type). or am I missing something obvious here?